tck/src/main/java/org/eclipse/microprofile/lra/tck/service/spi/LRARecoveryService.java

```diff
@@ -0,0 +1,85 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.eclipse.microprofile.lra.tck.service.spi;
+
+import java.net.URI;
+import java.util.logging.Logger;
+
+/**
+ *     The LRA model guarantees eventual consistency but does not say when
+ *     participants will be brought into a consistent state. When a request
+ *     to close/cancel an LRA is received, the implementation is allowed to
+ *     delay sending out the complete/compensate callbacks (the specification
+ *     only requires that an implementation must eventually send them).
+ *
+ *     This SPI provides a mechanism for an implementation to indicate when
+ *     it knows that callbacks have been sent and responses received or when
+ *     callbacks have been issued, all participants have reached an end state,
+ *     and when all listener notifications have been successfully delivered.
+ *
+ */
+public interface LRARecoveryService {
+
+    /**
+     * Wait for the delivery of Complete and Compensate participant callbacks.
+     * When this method returns the caller can be certain that the callbacks were
+     * sent and responses received (including error responses).
+     *
+     * @param lraId the LRA context
+     * @throws LRACallbackException the implementation was unable to determine whether
+     * or not the callbacks were received by all participants
+     */
+    void waitForCallbacks(URI lraId) throws LRACallbackException;
+
+    /**
+     * Wait for all participants to reach an end state and for all
+     * {@link org.eclipse.microprofile.lra.annotation.AfterLRA} notifications to be successfully delivered.
+     * 
+     * The default implementation iterates {@link LRARecoveryService#waitForEndPhaseReplay(URI)} until
+     * all participants reach a final state and for all listeners notifications to be successfully delivered.
+     *
+     * @param lraId the LRA context
+     * @throws LRACallbackException the implementation was unable to determine whether
+     * or not all participants have reached an end state and whether or not all listeners have been
+     * successfully notified
+     */
+    default void waitForRecovery(URI lraId) throws LRACallbackException {
+        Logger log = Logger.getLogger(LRARecoveryService.class.getName());
+        int counter = 0;
+
+        do {
+            log.info("Recovery attempt #" + ++counter);
+        } while (!waitForEndPhaseReplay(lraId));
+        log.info("LRA " + lraId + "has finished the recovery");
+    }
+
+    /**
+     * Wait for one replay of the end phase of the LRA (the calls to Status, Complete, Compensate, and Forget
```

Comment:
It is not entirely clear for me, also the difference between `waitForRecovery` and `waitForEndPhaseReplay`.

this `waitForEndPhaseReplay` issues the callbacks and wait for a response from them.  In which cases does it return false?

- When the callback returned an error (say status 500)
- When the callback returns an 'in progress' status (like ParticipantStatus.compensating)

So correct me if I'm wrong, but it should only return false in the case the end state is not yet achieved yet, but it makes sense to call this method again (which is what `waitForRecovery` does)

If it doesn't make sense to call the method again

- the callback returned an error (say status 500)
- the callback returned a failed state (ParticipantStatus.FailedToCompensate)
- ...

It must throw an exception.

Reply:
yes, you got everything correct. If exception is thrown this is unrecoverable state for the SPI (can by handled in TCK test), false means that one round of callbacks was issued but LRA is still not finished (202, or error from one or more of the callbacks) and true means that after the round of callbacks the LRA is in the end state (Closed, Cancelled, FailedTo*). Do you think we should adjust something in the interface to make it more clear?

Reply:
I would add the above examples to the javadoc so that each implementor knows exactly what it should do in each case. Probably most people will 'guess' it correctly, as I did, but it putting this explicitly in the text is probably easier for implementors.
Thx

Reply:
I hit send too soon.

> If it doesn't make sense to call the method again

> the callback returned an error (say status 500)

No, and I can see how this is misleading. We are testing this scenario in TckRecoveryTests so I would prefer to just return false in this case and keep the exception for really exceptional cases.

> the callback returned a failed state (ParticipantStatus.FailedToCompensate)

No, the LRA reached a final state (FailedToCancel) to it should return true